### PR TITLE
Fix #628 & #629 & #621: Store event log index for all records

### LINKED ISSUE #628: Store event log index for all records
For each event that is processed and stored into any entity, we need to store eventIndex as well.

Some entities are already using eventIndex when generating the id, but this is not enough.

Rationale:
 - we are working on having less txs ->  one tx could have multiple actions (like multiple DT buying,  etc) with multiple startOrder events
 - have multiple providerFees in one tx
 - have multiple ve+df actions in one tx
 - etc

Customers: D______ & F___:
- When they're doing compute jobs, they need to do several txs. It's a big pain for them
- Doing this issue will solve this (?)

### PR DESCRIPTION
**Pull request by mariacarmina**
Fixes #628, #629, #621

1. Additionally, updated with the latest version of ocean.js which contains the Multichain provider.
2. Implemented & tested handlers for `PublishMarketFee` and `ConsumeMarketFee` events.
3. Integrated search mechanisms for orders' & reused orders' IDs.
4. Generated entities' IDs with eventIndex wherever needed.
5. Updated `schema.graphql` entities with `eventIndex` field.
6. Fixed tests that include the modifications.

### CODE REVIEW

**Comment by alexcos20** [comment]
just two changes in schema ?  This is way to less...

To be easier for developer and reviewer , please make sure that in schema , anywhere you have txId, right after that line you will add eventIndex..

Something like: 
`
 tx: String!`
 
 will be replaced with :
 
 `
  tx: String!
  eventIndex: Int!`

**Comment by mihaisc** [comment]
In erc20template, in getOrderId also add event.params.serviceIndex.toI32() . Multiple services can be ordered in same tx

**Comment by mariacarmina** [comment]
I'll do #629 in this PR as well, since they are related and depend on each other.

**Comment by mariacarmina** [comment]
After attaching `eventIndex` to every entity that had `tx`, I started to add `eventIndex` in ID generation as the purpose of #629.
I have noticed that when `startOrder` is called, it generates 2 events: `OrderStarted` and `ProviderFee`. These 2 events have different handlers:

1. For `OrderStarted` event
```
handleOrderStarted(event: OrderStarted)
```

2. For `ProviderFee` event

```
handleProviderFee(event: ProviderFee)
```
Inside those handlers, `orderId` is processed by `getOrderId` which should receive additionally as a parameter, the order `eventIndex`. For (1) it works fine, because it will process first the ID. For (2) it won't work when fetching the existing order, because the handler for `ProviderFee` gets only this event. I need somehow to let the handler for provider fee to receive an event of `OrderStarted.`, otherwise order query will not work to retrieve the provider fees, because the `provider fees` are null after `startOrder` is called.

```console
queryResult:  {
  data: {
    order: {
      id: '0x25fa5236a3911b2d7262506a6580631494d99decf1ba293b470b084f226c68f8-0xa74847bf9cebd3696031dc8f0033cf9a009e66bc-0xa78deb2fa79463945c247991075e2a0e98ba7a09-0',
      providerFee: null,
      lastPriceToken: [Object],
      eventIndex: 0
    }
  }
}

```
This problem will show up to any fees event. I have thought of creating separate entities for every fee event and maybe somehow analyze deeper the transaction elements in order to figure out which order depend on which fees & so on. I assume that this will not be "queryable" at all and hard to use/query.

After discussing with @mihaisc, there is another approach of creating different templates of ERC20 which have the same functionality but with more details in the events. The disadvantage will be that subgraph will need to cover the existing templates and to do sort of migrations in order to migrate from the new templates to the old templates.

CC @alexcos20 @mihaisc

**Comment by alexcos20** [comment]
No need to go that far.
What we know:
   - each event in a tx has an unique eventIndex
   - subgraph processes events in a tx according to their order, so when handleProviderFee is called, you know that handleOrderStarted has already been process.
   
  Here is how to solve this:
 - for orders,  id =  txhash - eventId
- for providerFee, id =  txhash - eventId
 - in handleProviderFee, search for existing orders in a loop, substracting one from providerFee.eventIndex until you find an order entity that has order.datatoken.id = event.address  (so we know that is generated by the same dt contract).

In a nutshell:
  - given txHash 'abc',  you will end up with  order id 'abc-1' and providerFee id 'abc-2', linked to order 'abc-1'

**Comment by alexcos20** [comment]
Same approach can be used for PublishMarketFee and ConsumeMarketFee events

**Comment by mariacarmina** [comment]
I have fixed the tests for orders, integrating the event index into order IDs and implemented the search mechanism for order event indexes. Additionally I have implemented the handlers for `PublishMarketFee` & `ConsumeMarketFee` events and tested them in `Datatoken.test.ts`, so basically this PR also fixes #621.

**Comment by alexcos20** [comment]
on it, but it will take some time to review it

File: src/mappings/erc20Templates.ts [hunk 1]
```diff
@@ -235,40 +292,28 @@ export function handleProviderFee(event: ProviderFee): void {
...
+    event.address.toHex(),
+    event.logIndex.toI32()
+  )
   if (orderReuse) {
+    log.info('order reuse id in provider fee handler: {}', [orderReuse.id])
     orderReuse.providerFee = providerFee
     orderReuse.providerFeeValidUntil = event.params.validUntil
     orderReuse.save()
-  } else {
-    orderReuse = new OrderReuse(event.transaction.hash.toHex())
```
Comment (jamiehewitt15) [inline src/mappings/erc20Templates.ts:258]: Why did you delete this?

File: test/integration/Datatoken.test.ts [hunk 2]
```diff
@@ -169,7 +170,7 @@ describe('Datatoken tests', async () => {
       method: 'POST',
       body: JSON.stringify(initialQuery)
     })
-    await sleep(2000)
+    await sleep(3000)
```
Comment (jamiehewitt15) [inline test/integration/Datatoken.test.ts:173]: Just checking... is it really necessary to increase all of the sleep time to get this test to pass?

File: test/integration/Datatoken.test.ts [hunk 3]
```diff
@@ -214,6 +215,10 @@ describe('Datatoken tests', async () => {
     assert(dt.block >= blockNumber, 'incorrect value for: block')
     assert(dt.block < blockNumber + 50, 'incorrect value for: block')
     assert(dt.lastPriceValue === '0', 'incorrect value for: lastPriceValue')
+    assert(
+      dt.eventIndex !== null && dt.eventIndex > 0,
```
Comment (jamiehewitt15) [inline test/integration/Datatoken.test.ts:219]: Can we check for the exact value rather than just checking that it's above 0?

File: test/integration/Datatoken.test.ts [hunk 4]
```diff
@@ -325,10 +331,41 @@ describe('Datatoken tests', async () => {
...
```
Comment (jamiehewitt15) [inline test/integration/Datatoken.test.ts:335]: also here, can we check for the exact value of eventIndex?

File: src/mappings/erc20Templates.ts (hunk 1 above)
Comment (mariacarmina) [reply src/mappings/erc20Templates.ts:258]: In this case either `OrderStarted`, either `OrderReuse` must be fetched before `ProviderFee` event. They must already exist in subgraph.

cc @mihaisc

File: test/integration/Datatoken.test.ts (hunk 3 above)
Comment (mariacarmina) [reply test/integration/Datatoken.test.ts:219]: I can put the exact value, the reason that I put like that is when testing locally multiple times, that value differs.

File: test/integration/Datatoken.test.ts [hunk 5]
```diff
@@ -340,42 +377,76 @@ describe('Datatoken tests', async () => {
...
     })
-
+    await sleep(3000)
     const initialToken = (await response.json()).data.token
 
     assert(initialToken, 'Invalid initialToken')
     assert(initialToken.orderCount === '0', 'Invalid initial orderCount')
     assert(initialToken.orders.length === 0, 'Invalid initial orders')
+    assert(
+      initialToken.eventIndex !== null && initialToken.eventIndex > 0,
```
Comment (jamiehewitt15) [inline test/integration/Datatoken.test.ts:447]: Again, let's check the exact value

File: test/integration/SimplePublishConsume.test.ts [hunk 6]
```diff
@@ -287,18 +318,20 @@ describe('Simple Publish & consume test', async () => {
...
     }
 
-    await sleep(2000)
+    await sleep(3000)
     const response = await fetch(subgraphUrl, {
       method: 'POST',
       body: JSON.stringify(query)
     })
-
+    await sleep(3000)
```
Comment (jamiehewitt15) [inline test/integration/SimplePublishConsume.test.ts:334]: This isn't necessary. The `await` will already ensure that you get a response, the sleep doesn't change anything

**Comment by mariacarmina** [comment]
> Don't focus that much on improving the tests, obviously fix if there is an issue. This is taking long and we should merge it. Focus on the resolving the conflicts

Create a separate issue #665.

**Comment by jamiehewitt15** [comment]
We have conflicts on the PR now, could you resolve them?

**Review by jamiehewitt15 — APPROVED** [review]
The changes look good. There are some improvements we could do with the tests but those could be tackled separately (https://github.com/oceanprotocol/ocean-subgraph/issues/665)
